volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,777 @@
+
+from volatility3.framework import interfaces, exceptions
+from volatility3.framework import renderers, constants
+from volatility3.framework.configuration import requirements
+from volatility3.framework.renderers import format_hints
+from volatility3.framework.symbols import intermed
+from volatility3.framework.symbols.windows import pdbutil
+from volatility3.framework.symbols.windows.extensions import pe
+from volatility3.plugins.windows import pslist, modules
+
+vollog = logging.getLogger(__name__)
+
+
+class PESymbolFinder:
+    """
+    Interface for PE symbol finding classes
+    This interface provides a standard way for the calling code to
+    lookup symbols by name or address
+    """
+
+    cached_str = Union[str, None]
+    cached_str_dict = Dict[str, cached_str]
+
+    cached_int = Union[int, None]
+    cached_int_dict = Dict[str, cached_int]
+
+    cached_value = Union[int, str, None]
+    cached_value_dict = Dict[str, Union[Dict[str, List[str]], Dict[str, List[int]]]]
+
+    def __init__(
+        self,
+        layer_name: str,
+        mod_name: str,
+        module_start: int,
+        symbol_module: Union[interfaces.context.ModuleInterface, pefile.ExportDirData],
+    ):
+        self._layer_name = layer_name
+        self._mod_name = mod_name
+        self._module_start = module_start
+        self._symbol_module = symbol_module
+
+        self._address_cache: PESymbolFinder.cached_int_dict = {}
+        self._name_cache: PESymbolFinder.cached_str_dict = {}
+
+    def _get_cache_key(self, value: cached_value) -> str:
+        """
+        Maintain a cache for symbol lookups to avoid re-walking of PDB symbols or export tables
+        within the same module for the same address in the same layer
+        """
+        return f"{self._layer_name}|{self._mod_name}|{value}"
+
+    def get_name_for_address(self, address: int) -> cached_str:
+        cached_key = self._get_cache_key(address)
+        if cached_key not in self._name_cache:
+            name = self._do_get_name(address)
+            self._name_cache[cached_key] = name
+
+        return self._name_cache[cached_key]
+
+    def get_address_for_name(self, name: str) -> cached_int:
+        cached_key = self._get_cache_key(name)
+        if cached_key not in self._address_cache:
+            address = self._do_get_address(name)
+            self._address_cache[cached_key] = address
+
+        return self._address_cache[cached_key]
+
+    def _do_get_name(self, address: int) -> cached_str:
+        raise NotImplementedError("_do_get_name must be overwritten")
+
+    def _do_get_address(self, name: str) -> cached_int:
+        raise NotImplementedError("_do_get_address must be overwritten")
+
+
+class PDBSymbolFinder(PESymbolFinder):
+    """
+    PESymbolFinder implementation for  PDB modules
+    """
+
+    def _do_get_address(self, name: str) -> PESymbolFinder.cached_int:
+        try:
+            return self._symbol_module.get_absolute_symbol_address(name)
```

Comment:
Yes, but they are constructed explicitly in the class for the specific module type.

Reply:
Well, so why are they sharing an interface if you can't actually call them with the same parameters?  There's no point giving them the same clothes if you can easily tell the difference by looking at their face?

Reply:
It allows for the same interface for resolving name->address and address->name regardless of the way we find the symbols.

Reply:
But, it's not the same interface because they take different types, and we don't then use the interface at the inputs because of it.  They can have the same kind of interface, but trying to make them share the same interface and then not using it seems like throwing it in there for the sake of it?  If the two interfaces deviate it doesn't break any code or anything.

Reply:
Fixed